### PR TITLE
PE-1530: Add Custom MetaData to File Uploads

### PR DESCRIPTION
**Pull request by fedellen**
This PR applies the Custom MetaData feature added in ArDrive-Core-JS to the `upload-file` command.

See: https://github.com/ardriveapp/ardrive-core-js/pull/196

### CODE REVIEW

File: src/CLICommand/parameters_helper.ts [hunk 1]
```diff
@@ -323,3 +354,6 @@ export class ParametersHelper {
 		return !!dryRun;
 	}
 }
+
+const invalidSchemaError =
```
Comment (matibat) [inline src/CLICommand/parameters_helper.ts:358]: Consider renaming to `invalidSchemaErrorMessage` or similar

File: src/parameter_declarations.ts [hunk 2]
```diff
@@ -466,3 +469,9 @@ Parameter.declare({
 	description:
 		'(OPTIONAL) Provide a custom content type to all files within the upload to be used by the gateway to display the content'
 });
+
+Parameter.declare({
+	name: CustomTagsParameter,
+	aliases: ['--custom-tags'],
```
Comment (matibat) [inline src/parameter_declarations.ts:475]: Consider using `--custom-tags-file` or similar. We might end up implementing an extra `--custom-tags` of type array later on

**Review by matibat — COMMENTED** [review]
Looking good

File: src/commands/upload_file.ts [hunk 3]
```diff
@@ -55,14 +56,14 @@ function getFilesFromCSV(parameters: ParametersHelper): UploadPathParameter[] |
...
-		// eslint-disable-next-line prettier/prettier
+		// prettier-ignore
 		const [localFilePath, destinationFileName, _parentFolderId, drivePassword, _driveKey, _customContentType] =
-			// eslint-disable-next-line prettier/prettier
 			csvFields;
 
 		const customContentType = _customContentType ?? parameters.getParameterValue(CustomContentTypeParameter);
 
 		// TODO: Make CSV uploads more bulk performant
+		// TODO: Make CSV work with custom metadata -- Question: How to apply this? Do we only allow metadata JSON file paths for CSV? Do we accept a stringified JSON object directly in the CSV?
```
Comment (fedellen) [inline src/commands/upload_file.ts:66]: We should have a ticket that applies custom metadata for CSV

File: src/CLICommand/parameters_helper.ts [hunk 4]
```diff
@@ -282,6 +289,84 @@ export class ParametersHelper {
...
+			}
+			if (isCustomMetaDataTagInterface(tagJSON)) {
+				// User submits an object of names and values with no instructions on where to put them
+				// By default, We will put them on the File MetaData JSON
+				return { metaDataJson: tagJSON };
+			}
+		} catch {
+			throw Error(invalidSchemaError);
+		}
+		throw Error(invalidSchemaError);
```
[the remark this answers is not in the document]
Comment (fedellen) [reply src/CLICommand/parameters_helper.ts]: Applied at 044278b (CLI) 080651c (Core)

File: src/CLICommand/parameters_helper.ts [hunk 5]
```diff
@@ -282,6 +289,84 @@ export class ParametersHelper {
...
+		if (metadata === undefined || metadata.length === 0) {
+			return undefined;
+		}
+		this.assertEvenTags(metadata);
+
+		const metaData: CustomMetaDataTagInterface = {};
+		let temp: string | null = null;
+
+		for (const val of metadata) {
+			this.assertCharacterLength(val);
```
[the remark this answers is not in the document]
Comment (fedellen) [reply src/CLICommand/parameters_helper.ts]: Applied at https://github.com/ardriveapp/ardrive-cli/commit/044278b4f7b54c70a842db49896a831e721c994b (CLI) 080651c (Core)

File: src/parameter_declarations.ts (hunk 2 above)
Comment (fedellen) [reply src/parameter_declarations.ts]: Changed this to `--metadata-file` and `--metadata-json` + `metadata-gql-tags`

File: src/CLICommand/parameters_helper.ts (hunk 1 above)
Comment (fedellen) [reply src/CLICommand/parameters_helper.ts]: Renamed in Core: 080651c (Core)

**Comment by fedellen** [comment]
> We have a conflict in the `README.md`.

merged here: https://github.com/ardriveapp/ardrive-cli/pull/291/commits/443f2eb98eb3b0dd93f5772f1216fa9b9beb3243

**Review by matibat — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -1162,6 +1163,94 @@ https://arweave.net/{dataTxId}/custom-file-1
...
+
+When the custom metadata is attached to the MetaData Transaction's GQL tags, they will become visible on any Arweave GQL gateway and also third party tools that read GQL data.
+
+When these tags are added to the MetaData Transaction's Data JSON they can be read by downloading the JSON data directly from `https://arweave.net/METADATA_TX_ID`.
+
+To add this custom metadata to your file metadata transactions, CLI users can pass custom metadata with three different parameters:
+
+-   `--metadata-file`
+-   `--metadata-json`
+-   `--metadata-gql-tags`
```
Comment (fedellen) [inline README.md:1192]: Add some more info about the separation of these params

Comment (fedellen) [reply README.md:1192]: Optional and combinable

File: src/CLICommand/parameters_helper.ts [hunk 7]
```diff
@@ -282,6 +290,72 @@ export class ParametersHelper {
...
+
+		if (isCustomMetaData(tagJSON)) {
+			// Valid custom metadata schema we can send to core
+			return tagJSON;
+		}
+		if (assertCustomMetaDataTagInterface(tagJSON)) {
+			// User submits an object of names and values with no instructions on where to put them
+			// By default, We will put them on the File MetaData JSON
+			return { metaDataJson: tagJSON };
+		}
```
Comment (fedellen) [inline src/CLICommand/parameters_helper.ts:325]: Excise convenience flow

File: README.md [hunk 8]
```diff
@@ -1162,6 +1163,94 @@ https://arweave.net/{dataTxId}/custom-file-1
...
 ```
 
+### Uploading Files with Custom MetaData
+
+ArDrive CLI has the capability of attaching custom metadata to ArFS File and Folder MetaData Transactions during the `upload-file` command. This metadata can be applied to either the GQL tags on the MetaData Transaction and/or into the MetaData Transaction's Data JSON.
+
+All custom metadata applied must ultimately adhere to the following JSON shape:
+
+```ts
+type CustomMetaDataTagInterface = Record<string, string | string[]>;
```
Comment (karlprieb) [inline README.md:1173]: We need to update this

File: README.md [hunk 9]
```diff
@@ -1162,6 +1163,94 @@ https://arweave.net/{dataTxId}/custom-file-1
...
+    "metaDataJson": {
+        "Tag-Name": ["Value-1", "Value-2"]
+    },
+    "metaDataGqlTags": {
+        "GQL Tag Name": "Tag Value"
+    }
+}
+```
+
+The `--metadata-json` and `--metadata-gql-tags` parameters accept an array of string values to be applied to their respective target (MetaData Tx Data JSON or GQL Tags). This method of CLI input does not support multiple tag values for a given tag name and the input must be an EVEN number of string values. e.g:
```
Comment (karlprieb) [inline README.md:1219]: update "accept an array of string values"

File: src/CLICommand/parameters_helper.ts (hunk 7 above)
Comment (fedellen) [reply src/CLICommand/parameters_helper.ts]: https://github.com/ardriveapp/ardrive-cli/pull/291/commits/8ea7dc0935cc35459a504b35a61775067284b55a

File: README.md (hunk 8 above)
Comment (fedellen) [reply README.md]: a8d367e

File: README.md (hunk 6 above)
Comment (fedellen) [reply README.md]: Added how the param is used for less confusion and clear separation: a8d367e

File: README.md [hunk 10]
```diff
@@ -1162,6 +1163,100 @@ https://arweave.net/{dataTxId}/custom-file-1
...
+export type JsonSerializable =
+    | string
+    | number
+    | boolean
+    | null
+    | { [member: string]: JsonSerializable }
+    | JsonSerializable[];
+```
+
+e.g:
```
Comment (matibat) [inline README.md:1188]: Nit (although I'm not sure)
```suggestion
e.g.:
```

**Review by matibat — APPROVED** [review]
LGTM

**Review by karlprieb — APPROVED** [review]

**Review by arielmelendez — APPROVED** [review]

**Review by matibat — APPROVED** [review]
LGTM!

Just tested it for bulk upload: custom metadata gql tags and custom metadata json (not metadata file, which was previously working).

**Review by matibat — CHANGES_REQUESTED** [review]
Requirements just changed. We are gonna delay the release of CLI.

**Review by matibat — APPROVED** [review]
CLI with CORE v1.16.0 ready to go!

**Pull request merged**